bootstrap/src/main/java/com/splunk/opentelemetry/javaagent/bootstrap/metrics/jmx/JmxMetricsWatcher.java

```diff
@@ -0,0 +1,193 @@
+    this.query = query;
+    this.metersFactory = metersFactory;
+  }
+
+  // copied from micrometer CommonsObjectPool2Metrics class
+  private static MBeanServer findMBeanServer() {
+    List<MBeanServer> mBeanServers = MBeanServerFactory.findMBeanServer(null);
+    if (!mBeanServers.isEmpty()) {
+      return mBeanServers.get(0);
+    }
+    return ManagementFactory.getPlatformMBeanServer();
+  }
+
+  public void start() {
+    try {
+      // get all mbeans matching the query, register meters for them
+      Set<ObjectName> existingObjects = mBeanServer.queryNames(query.toObjectNameQuery(), null);
+      for (ObjectName existingObject : existingObjects) {
+        meters.compute(existingObject, this::computeMeters);
+      }
+
+      // register a notification listener that'll add and remove meters as they come
+      NotificationFilter filter = this::isNotificationEnabled;
+      notificationListener = this::handleNotification;
+      mBeanServer.addNotificationListener(
+          MBeanServerDelegate.DELEGATE_NAME, notificationListener, filter, null);
+    } catch (JMException e) {
+      log.warn("Could not start JMX metrics listener", e);
+    }
+  }
+
+  public void stop() {
+    if (notificationListener != null) {
+      try {
+        mBeanServer.removeNotificationListener(
+            MBeanServerDelegate.DELEGATE_NAME, notificationListener);
+        notificationListener = null;
+      } catch (InstanceNotFoundException | ListenerNotFoundException e) {
+        log.warn("Could not remove JMX metrics listener from the MBeanServer", e);
+      }
+    }
+
+    executorService.shutdown();
+
+    meters.values().stream().flatMap(Collection::stream).forEach(meterRegistry::remove);
+    meters.clear();
+  }
+
+  // visible for tests
+  Set<Meter> getMeters() {
+    return meters.values().stream().flatMap(Collection::stream).collect(toSet());
+  }
+
+  private List<Meter> computeMeters(ObjectName objectName, List<Meter> meters) {
+    if (meters == null) {
+      meters = new CopyOnWriteArrayList<>();
+    }
+    try {
+      meters.addAll(metersFactory.createMeters(mBeanServer, objectName));
+    } catch (JMException e) {
+      throw new IllegalStateException(
+          "Could not create meters for JMX object; most likely a programming error", e);
+    }
+    return meters;
+  }
+
+  private boolean isNotificationEnabled(Notification notification) {
+    if (!handledNotificationTypes.contains(notification.getType())) {
+      return false;
+    }
+    ObjectName objectName = ((MBeanServerNotification) notification).getMBeanName();
+    return query.matches(objectName);
+  }
+
+  private void handleNotification(Notification notification, Object handback) {
+    // we can't get JMX object attributes in the NotificationListener, so we'll do that
+    // asynchronously on a separate thread
+    executorService.submit(() -> handleNotificationAsync(notification));
+  }
+
+  private void handleNotificationAsync(Notification notification) {
```

Comment:
I think the names of `handleNotification` and `handleNotificationAsync` are swapped. To me, the async method should be the non-blocking one and the other one should be the one that does the work (blocking).

Reply:
That's totally true 🙈  It seems I mixed them up and didn't notice - thanks!